GrafanaUI: Add success state to ClipboardButton

evictorero opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
This PR starts as a suggestion from UX that in the `ClipboardButton` the feedback of the action is too far from where the action is happening in Public Dashboards.
We are showing a toast notification at the upper right corner of the screen every time a user copies a link to a clipboard.
We collect some feedback from other teams and it seems we do not have a convention for doing this so this PR is an attempt to standardize this functionality across all Grafana.

Changes inside the `ClipboardButton` component
* Added feedback inside the button changing the text by `Copied!`
* Added feedback inside the button changing icon by `check` along the text

Changes from the usages of ClipboardButton
* Removed all appNotifications (toasts) from the callers
* Changed color to primary for most cases when it makes sense. Maybe we should do it in all of them?
* Removed nested Icon component and use the property with 'copy' icon to improve UI padding
* In the Snapshot sharing tab: 
    * Changed link and color button in favor of input and blue button for consistency
**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes https://github.com/grafana/grafana-enterprise-partnerships-team/issues/269

**Special notes for your reviewer**:
Is it necessary to be explicit about “copy to clipboard” or we should change all of those by just “copy”?
I think is not ideal to change the size of the button after clicking it but this PR could be the start of consolidating all uses into something. We could iterate this.

I put some screenshots [in this document](https://docs.google.com/document/d/1KucZf7gEAh_UGrThxjscAbjtRmWCpnlgoC8tflj-b68/edit?usp=sharing) with the old and new UI so it is easier to see and compare the differences.

evictorero committed:
User Experience: apply the same pattern feedback for all copy to clipboard buttons

evictorero committed:
Merge branch 'main' into evictorero/copy-button-feedback

evictorero committed:
add copy icon to all ClipboardButton use cases

evictorero committed:
Change primary color for copy to clipboard in create token

owensmallwood reviewed (approved):
Nice 🎉

joshhunt commented on packages/grafana-ui/src/components/Table/TableCellInspectModal.tsx:
```diff
@@ -60,15 +59,8 @@ export function TableCellInspectModal({ value, onDismiss, mode }: TableCellInspe
-              Copied to clipboard
-            </>
-          )}
+        <ClipboardButton icon="check" getText={() => text} onClipboardCopy={() => setIsInClipboard(true)}>
```
I don't think this needs the `onClipboardCopy={() => setIsInClipboard(true)}` and more (and the associated code from it) as that was doing the same "Copied!" stuff that you've added in the base component

joshhunt reviewed (commented):
Great stuff, very thorough! 

I'm not sure if the interaction is the best, with the button changing size after copied. Will discuss this in daily sync and see what we want to do (especially within the scope of _your_ PR)

joshhunt commented on packages/grafana-ui/src/components/ClipboardButton/ClipboardButton.tsx:
```diff
@@ -12,9 +12,16 @@ export interface Props extends ButtonProps {
+    setButtonIcon('check');
 
     try {
       await copyText(textToCopy, buttonRef);
```
Also, we should set the success state once we actually successfully copy text - after the `await copyText` call!

joshhunt committed:
Add success button variant

joshhunt committed:
Remove copy confirmation from TableCellInspectModal because it's in the base component now

joshhunt committed:
Design tweaks to copy confirmation

 - Only change the icon to tick to avoid the button changing size
 - Change button to success green
 - Only show copy confirmation state for 2 seconds

joshhunt committed:
Merge remote-tracking branch 'origin/main' into evictorero/copy-button-feedback

joshhunt committed:
revert TabelCellInspectModal text button back

ivanortegaalba commented on public/app/features/dashboard/components/ShareModal/ShareLink.tsx:
```diff
@@ -99,7 +94,7 @@ export class ShareLink extends PureComponent<Props, State> {
           <Field
             label="Lock time range"
-            description={isRelativeTime ? 'Transforms the current relative time range to an absolute time range' : ''}
+            description={isRelativeTime && 'Transforms the current relative time range to an absolute time range'}
```
Could this be rendered `"null"` or `"false"` strings if `isRelativeTime` is a falsy value?

joshhunt replied:
this actually shouldn't be here at all!!!!!! unrelated change i was testing out (to answer your exact question 😅

joshhunt committed:
revert accidental change to ShareLink

dprokop commented on packages/grafana-ui/src/components/Table/TableCellInspectModal.tsx:
```diff
@@ -60,15 +42,8 @@ export function TableCellInspectModal({ value, onDismiss, mode }: TableCellInspe
-              Copied to clipboard
-            </>
-          )}
+        <ClipboardButton icon="copy" getText={() => text}>
+          Copy to Clipboard
```
Thanks for updating this piece of code!

dprokop reviewed (approved):
Works and looks great code-wise! 

My only comment is around the copy of the button when the user clicked it. I wouldn't be too strict about not changing the width of the button due to the copy change. IMO as long as the button doesn't change its position, subtle width change should be just fine.
